### ISSUE TITLE
edge_index not defined

### ISSUE THREAD

**Issue by alemelis** [JuliaGraphs/LightGraphs.jl#698]
Hi, 
I'm rewriting part of my numerical library to support LigthGraphs.jl instead of [Graphs.jl](https://github.com/JuliaArchive/Graphs.jl). This is only because of the compatibility with julia > 0.5.

I am missing the [edge_index function](http://graphsjl-docs.readthedocs.io/en/latest/interface.html?highlight=edge_index#edge_index)

> edge_index(e, g)
returns the index of an edge e in graph g. Each edge must have a unique index between 1 and num_edges.

Is there something similar in LightGraphs?

**Comment by sbromberger** [comment]
Hi @alemelis - there is no `edge_index` because neither edges nor vertices are indexed. That is, you can't reference an edge in a LightGraph outside its iterator.

Can you tell me what you're trying to accomplish? Perhaps we can work out another way to do it.

**Comment by alemelis** [comment]
Hi @sbromberger , I am using Graphs.jl to store store and navigate a graph which represents the topology of a vascular network (something like [this](https://insigneo.github.io/openBF/images/grafo.png)). Each edge in the graph represents an artery and my library ([openBF](https://github.com/INSIGNEO/openBF/tree/julia6)) is then used to solve PDE describing the blood flow.

An example of how Graphs.jl is used is [here](https://insigneo.github.io/openBF/Docs/doc-files/grafo.html). 

You can also see the actual implementation within openBF [here](https://github.com/INSIGNEO/openBF/blob/julia6/src/godunov.jl#L552).

In short, I fill a `LightGraphs.Graph()` structure through `LightGraphs.add_edge!()` function.
For each iteration of the main cycle, the list of edges is read and I check whether the edge has 0, 1, or 2 other edges connected to either the source or destination vertices (`LightGraphs.indegree()` or `LightGraphs.outdegree()`, respectively). Finally, I want to know which are the edges connected. This was done with `edge_index`.

**Comment by sbromberger** [comment]
Can you point to the code in openBF that creates and then uses the Graph? I'm pretty sure we can do this.

Note that edges are immutable, so knowing which edges are connected merely implies Edge A == Edge B.

**Comment by alemelis** [comment]
The graph is created in the [`main.jl` file](https://github.com/INSIGNEO/openBF/blob/1c787138acd8bfbeee485db664fd967ed9e3befb/test/single-artery/main.jl#L106-L121) and then [passed](https://github.com/INSIGNEO/openBF/blob/1c787138acd8bfbeee485db664fd967ed9e3befb/test/single-artery/main.jl#L198) to the `solveModel()` function in [godunov.jl](https://github.com/INSIGNEO/openBF/blob/1c787138acd8bfbeee485db664fd967ed9e3befb/src/godunov.jl#L520-L570) where the index reference is needed.

**Comment by sbromberger** [comment]
Couple of comments:

1) why are you passing the edge iterator (`edgess`) along with the graph? I don't think you need that, since the edge iterator is already contained within the graph.

2) there are a few places where I'm having a problem understanding your code. For example: https://github.com/INSIGNEO/openBF/blob/1c787138acd8bfbeee485db664fd967ed9e3befb/src/godunov.jl#L549-L552:

what is `joinVessels` doing here? In general wherever you have `edge_index`, you should consider just passing the edge: in this case, `Edge(t, o[1])`.

This will require `vessels` to be a dictionary indexed by `Edge` instead of a vector of Ints.
```
elseif LightGraphs.outdegree(grafo, t) == 1
      if length(LightGraphs.in_neighbors(grafo, t)) == 1
        o = LightGraphs.out_neighbors(grafo, t)
        openBF.joinVessels(blood, vessels[i], vessels[LightGraphs.edge_index(o[1], grafo)])
```

However: if you're creating the graph *exactly once* and not changing it, then you could theoretically do
```
edgelist = collect(edges(grafo))

```
and since you're not mutating the graph, the edgelist should never change. Then,
```
edgemap = Dict{Edge, Int}(e=>i for (e, i) in enumerate(edgelist))
edge_index(e::Edge) = edgemap(e)
```

This is fine as long as you're not worried about memory usage (if you have extremely large graphs, this is probably not what you want to do, but anything < ~10^6 edges or so should be fine).

**Comment by alemelis** [comment]
> why are you passing the edge iterator (edgess) along with the graph? I don't think you need that, since the edge iterator is already contained within the graph.

Well spotted, indeed there is not reason to pass both the graph and the iterator.

> what is joinVessels doing here?

[`joinVessels`](https://insigneo.github.io/openBF/Docs/doc-files/junctions.html#joinVessels) computes the numerical solution depending on how many edges are connected to a specific vertex. It takes as argument a list of `::vessel` type items (stored in the list `vessels` which is populated in the [main file](https://github.com/INSIGNEO/openBF/blob/1c787138acd8bfbeee485db664fd967ed9e3befb/test/single-artery/main.jl#L114)).

The `vessels` list is populated with information stored in a `.csv` file like 

```csv
name, sn, tn, other values...,
v1,   1,  2,  ...,
v2,   2,  3,  ...,
v3,   2,  4,  ...,
v4,   4,  5,  ...,
v5,   3,  6,  ..., 
```

where `sn` and `tn` are the source and destination vertex indices, and each line represents a `vessel` (and therefore an edge in the graph). The reason why I need to find the edge index is because reading the list of edges, not all the junctions can be found. For example, the graph from the csv above would be 

```
1 ---- 2 ---- 3 ----- 6
       |
       4 ---- 5
```

The first junction would start at vertex 2 involving vertices 1, 2, 3, and 4. The second junction is in vertex 4. The third junction is in vertex 3 and the destination vertex is the number 6.

I'll try the dictionary approach. The size is not a problem as in the worst case scenario the network would include few hundreds of edges.

**Comment by sbromberger** [comment]
> The size is not a problem as in the worst case scenario the network would include few hundreds of edges.

This shouldn't be an issue at all, and is probably the easiest way to transition to LightGraphs.

**Comment by alemelis** [comment]
Thank you so much for your help @sbromberger, the `edge_map` works like a charm

**Comment by sbromberger** [comment]
Glad I could help!
